README.md
<div align="center">
<p align="center">
    <img src="gretel-logo.png?raw=true?" alt="gretel-logo" width="200">
</p>
<h1 align="center">Gretel</h1>
<h3 align="center">An algorithm for recovering haplotypes from metagenomes. Sister to <a href="https://github.com/SamStudio8/hansel">Hansel</a>.
</h3>
<p align="center">
<a href="https://github.com/samstudio8/gretel/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-MIT-orange.svg" alt="License"></a>
<a href="https://bioconda.github.io/recipes/gretel/README.html"><img src="https://anaconda.org/bioconda/gretel/badges/downloads.svg" alt="bioconda"></a>
</p>
</div>


What is it?
-----------

**Gretel** is a Python package providing a command line tool for the recovery of haplotypes
from metagenomic data sets. **Gretel** parses an alignment of reads into a **Hansel** matrix
and uses the evidence of SNP pairs observed to appear on the same reads to probabilistically
reconstruct the most likely haplotypes.

**Gretel** uses an L'th order Markov chain model to reconstruct likely sequences
of variants that constitute haplotypes in the real metagenome.
Our approach involves graph-like traversal of the data within the **Hansel** matrix.
Edges are probabilitically weighted based on the evidence on the reads, as well as
the haplotype as it has been reconstructed so far.

What can I use it for?
----------------------

**Gretel** is designed to recover haplotypes from your data set, without the need for
setting (or optimisation) of any parameters.
**Gretel** does not require a priori knowledge of your input data (such as its contents, or
the true number of haplotypes) and makes no assumptions
regarding the distributions of alleles at variant sites and uses the available evidence
from the aligned reads without altering or discarding the observed varations.

Why should I use it?
--------------------

**Gretel** is the first tool capable of recovering haplotypes from metagenomes.
Whilst tools exist for analogous haplotyping problems, such as the assembly of
viral quasispecies, typically these tools rely on overlap approaches that create
too many unranked haplotypes. **Gretel** is capable of ranking the haplotypes it
outputs by their likelihood.

**Gretel** requires no parameters and our approach is robust to sequencing error
and misalignment noise.

Requirements
------------


    $ pip install numpy hanselx pysam PyVCF

Install
-------


    $ pip install gretel

Alternatively, Gretel has been packaged for bioconda (Thanks [@johnne!](https://github.com/johnne)):

    $ conda install -c bioconda gretel

Usage
-----
You will require a sorted BAM containing your reads, aligned to some pseudo-reference.
You can use any sequence as your reference, such as a consensus assembly of the
metagenomic reads, or a known strain reference (such as HIV-1).
You must bgzip and tabix your VCF.

    $ gretel <bam> <vcf.gz> <contig> -s <1-start> -e <1-end> --master <master.fa> -o <outdir>


Citation (Pre-Print)
--------
```
@article {Nicholls223404,
        author = {Nicholls, Samuel M. and Aubrey, Wayne and Edwards, Arwyn and Grave, Kurt de and Huws, Sharon and Schietgat, Leander and Soares, Andr{\'e} and Creevey, Christopher J. and Clare, Amanda},
        title = {Recovery of gene haplotypes from a metagenome},
        elocation-id = {223404},
        year = {2019},
        doi = {10.1101/223404},
        publisher = {Cold Spring Harbor Laboratory},
        URL = {https://www.biorxiv.org/content/early/2019/10/17/223404},
        eprint = {https://www.biorxiv.org/content/early/2019/10/17/223404.full.pdf},
        journal = {bioRxiv}
}
```

License
-------
Hansel and Gretel are distributed under the MIT license, see LICENSE.


CHANGELOG.rst
History
=======

0.0.94
------
* Added `--pepper` option to for permissive pileups by overriding the pysam pileup stepper to `all` instead of `samtools`.

0.0.93
------

* Move `process_vcf` to `util` module. I may drop use of `pyvcf` in future as I don't like the API.
* Dropped pointless `append_path` stub.
* Fixed an edge case where reads beginning with a SNP that aligned to the start of a parallel parsing window are counted twice.
* Added a small test package to help detect future regressions.
* Added `--version` argument to print program version number.
* Removed `--lorder` argument as users should not need to select the chain order.

0.0.92
------

* Adds `--dumpmatrix` and `--dumpsnps` debugging options.
* Clean up Hansel matrix initialisation.
* Add `gretel-snpper` command for generating naive VCF.
* Fix a regression where the `L` parameter of the matrix is incorrectly left unset.

0.0.90
------
Resolves a bug whereby SNPs are incorrectly parsed from the BAM if either:
    * its quality score is below 13
    * the read is overlapped by its primary mate

Well covered data sets need not be overly affected by the additional noise that
may have been introduced, but the problem is more noticeable with low coverage
and you may wish to reapply Gretel to affected data. Sorry.


0.0.81
------
* Add warning and advice when an entry in Hansel is missing evidence.
* Make the 'Unable to select' warning sound much less bad because it is normal.

0.0.8
-----
* Docs
* Deprecate `gretel-crumbs` command

0.0.7
-----
* Further improvements to parallel read processing
* Add `-` symbol to enable support for deletions

0.0.6b
------
* Fix setting of `L` parameter

0.0.6
-----
* MULTIPROCESSING
* Re-write read handling, again

0.0.5
-----
* `-s` and `-e` introduced to allow specification of positions between which
  to recover haplotypes
* Attempt some basic indel handling
* Fix a bug where the master sequence was altered by the output of each
  reported haplotype

0.0.4
-----
* Add experimental `--sentinels` option
* Improve docs

0.0.3
-----
* Hansel is now seperate from Gretel
* [Hansel] `get_marginal_at` is `now get_counts_at`
* [Hansel] `selext_next_edge_at` deprecated
* Gene recovery and likelihood plots are now on seperate panels
* Re-write methods to add observations to matrix to be less awful to read
* Drop `--hit` and `--gene` options to verification
* Replace verification script to `gretel-crumbs` command

0.0.2
-----
* Improve documentation.
* Provide `util` subpackage for filling `Hansel` structure with BAM observations.
* Explicitly provide possible symbols to `Hansel`.
* Improve plotting
* Remove `process_hits` and `process_refs` as these are no longer needed.
* Rename `establish_path` to `generate_path`
* Rename `add_ignore_support3` to `reweight_hansel_from_graph` so we have some sort of indication of what it does.
* Altered Sphinx configuation.

0.0.1
-----
* Import repository from `claw`.


tests/test_test.py
import unittest

EXP = [1,2,10]
def load_vcf(path, contig, start_pos, end_pos):
    from gretel import util
    return util.process_vcf(path, contig, start_pos, end_pos)

def load_bam(path, vcf_path, contig, start_pos, end_pos, threads):
    from gretel import util
    return util.load_from_bam(path, contig, start_pos, end_pos, load_vcf(vcf_path, contig, start_pos, end_pos), n_threads=threads)

class BasicRegressionTest(unittest.TestCase):

    def test_test(self):
        self.assertTrue(True)

    def test_vcf(self):
        VCF_h = load_vcf('tests/data/test.vcf.gz', 'hoot', 1, 19)
        self.assertEqual(VCF_h["N"], 3)
        self.assertEqual(VCF_h["snp_rev"], {0:1, 1:2, 2:10})
        self.assertEqual(VCF_h["snp_fwd"], {1:0, 2:1, 10:2})

        self.assertEqual(len(VCF_h["region"]), 19+1)
        for i in range(len(VCF_h["region"])):
            if i in EXP:
                self.assertEqual(VCF_h["region"][i], 1)
            else:
                self.assertEqual(VCF_h["region"][i], 0)
        return VCF_h

    def test_bam(self):

        for thread in [1,2]:
            hansel = load_bam('tests/data/test.bam', 'tests/data/test.vcf.gz', 'hoot', 1, 20, thread)

            self.assertEqual(hansel.n_slices, 5) # n reads
            self.assertEqual(hansel.n_crumbs, 9) # n snp connections
            self.assertTrue(hansel.L > 0) # Check L is actually set

            # Test a couple of elements in the Hansel matrix
            self.assertEqual(hansel.get_observation('_', 'A', 0, 1), 1)
            self.assertEqual(hansel.get_observation('A', 'A', 1, 2), 1)
            self.assertEqual(hansel.get_observation('A', 'A', 1, 3), 1)
            self.assertEqual(hansel.get_observation('A', 'A', 1, 4), 0)
            self.assertEqual(hansel.get_observation('C', 'C', 1, 2), 1)
            self.assertEqual(hansel.get_observation('C', 'C', 1, 3), 1)
            self.assertEqual(hansel.get_observation('C', 'C', 1, 4), 0)
            self.assertEqual(hansel.get_observation('T', 'T', 1, 2), 2)
            self.assertEqual(hansel.get_observation('G', 'G', 1, 2), 0)
            self.assertEqual(hansel.get_observation('G', 'G', 2, 3), 0)
            self.assertEqual(hansel.get_observation('G', 'G', 3, 4), 1)
            self.assertEqual(hansel.get_observation('G', '_', 4, 5), 1)


if __name__ == '__main__':
    unittest.main()


gretel/util.py
import pysam
import numpy as np
from math import ceil
from hansel import Hansel
import vcf

from multiprocessing import Process, Queue, Value
import sys

def get_ref_len_from_bam(bam_path, target_contig):
    """
    Fetch the length of a given reference sequence from a :py:class:`pysam.AlignmentFile`.

    Parameters
    ----------
    bam_path : str
        Path to the BAM alignment

    target_contig : str
        The name of the contig for which to recover haplotypes.

    Returns
    -------
    end_pos : int
        The 1-indexed genomic position at which to stop considering variants.
    """
    bam = pysam.AlignmentFile(bam_path)
    end = bam.lengths[bam.get_tid(target_contig)]
    bam.close()

    return end

def load_from_bam(bam_path, target_contig, start_pos, end_pos, vcf_handler, use_end_sentinels=False, n_threads=1, debug_reads=False, debug_pos=False, stepper="samtools"):
    """
    Load variants observed in a :py:class:`pysam.AlignmentFile` to
    an instance of :py:class:`hansel.hansel.Hansel`.

    Parameters
    ----------
    bam_path : str
        Path to the BAM alignment

    target_contig : str
        The name of the contig for which to recover haplotypes.

    start_pos : int
        The 1-indexed genomic position from which to begin considering variants.

    end_pos : int
        The 1-indexed genomic position at which to stop considering variants.

    vcf_handler : dict{str, any}
        Variant metadata, as provided by :py:func:`gretel.gretel.process_vcf`.

    use_end_sentinels : boolean, optional(default=False)
        Whether or not to append an additional pairwise observation between
        the final variant on a read towards a sentinel.

        .. note:: Experimental
          This feature is for testing purposes, currently it is recommended
          that the flag be left at the default of `False`. However, some
          data sets report minor performance improvements for some haplotypes
          when set to `True`.
          This flag may be removed at any time without warning.

    n_threads : int, optional(default=1)
        Number of threads to spawn for reading the BAM

    debug_reads : list{str}, optional
        A list of read names for which to print out debugging information

    debug_pos : list{int}, optional
        A list of positions for which to print out debugging information

    stepper : str, optional(default=samtools)
        The pysam pileup stepper to use

    Returns
    -------
    Hansel : :py:class:`hansel.hansel.Hansel`
    """

    hansel = Hansel.init_matrix(['A', 'C', 'G', 'T', 'N', "-", "_"], ['N', "_"], vcf_handler["N"])

    if not debug_reads:
        debug_reads = set([])
    if not debug_pos:
        debug_pos = set([])

    import random
    def progress_worker(progress_q, n_workers, slices, total_snps, crumbs):
        worker_pos = []
        worker_done = []
        for _ in range(0, n_workers):
            worker_pos.append(0)
            worker_done.append(0)

        while sum(worker_done) < n_workers:
            work_block = progress_q.get()
            worker_pos[work_block["worker_i"]] = work_block["pos"]
            if work_block["pos"] is None:
                worker_done[work_block["worker_i"]] = 1

                crumbs.value += work_block["crumbs"]
                slices.value += work_block["slices"]
                total_snps.value += work_block["covered_snps"]
                sys.stderr.write("%s\n" % ([ worker_pos[i] if status != 1 else None for (i, status) in enumerate(worker_done)]))
            if random.random() < 0.1:
                sys.stderr.write("%s\n" % ([ worker_pos[i] if status != 1 else None for (i, status) in enumerate(worker_done)]))
        return (slices, total_snps, crumbs)

    def bam_worker(bam_q, progress_q, worker_i):

        worker = worker_i

        slices = 0
        crumbs = 0
        covered_snps = 0

        bam = pysam.AlignmentFile(bam_path)

        while True:
            work_block = bam_q.get()
            if work_block is None:
                progress_q.put({
                    "pos": None,
                    "worker_i": worker_i,
                    "slices": slices,
                    "crumbs": crumbs,
                    "covered_snps": covered_snps,
                })
                break

            reads = {}
            dreads = set([])

            for p_col in bam.pileup(reference=target_contig, start=work_block["start"]-1, stop=work_block["end"], ignore_overlaps=False, min_base_quality=0, stepper=stepper):

                if p_col.reference_pos + 1 > end_pos:
                    # Ignore positions beyond the end_pos
                    break

                if vcf_handler["region"][p_col.reference_pos+1] != 1:
                    # Ignore non-SNPs
                    continue

                for p_read in p_col.pileups:

                    curr_read_1or2 = 0
                    if p_read.alignment.is_paired:
                        if p_read.alignment.is_read1:
                            curr_read_1or2 = 1
                        elif p_read.alignment.is_read2:
                            curr_read_1or2 = 2
                        else:
                            #TODO Probably indicative of bad data
                            pass


                    curr_read_name = "%s_%s_%d" % (p_read.alignment.query_name, str(p_read.alignment.flag), curr_read_1or2)

                    LEFTMOST_1pos = p_read.alignment.reference_start + 1 # Convert 0-based reference_start to 1-based position (to match region array and 1-based VCF)

                    # Special case: Consider reads that begin before the start_pos, but overlap the 0th block
                    if work_block["i"] == 0:
                        if LEFTMOST_1pos < start_pos:
                            # Read starts before the start_pos
                            if p_read.alignment.reference_start + 1 + p_read.alignment.query_alignment_length < start_pos:
                                # Read ends before the start_pos
                                continue
                            LEFTMOST_1pos = start_pos
                    else:
                        # This read begins before the start of the current (non-0) block
                        # and will have already been covered by the block that preceded it
                        if LEFTMOST_1pos < work_block["start"]:
                            continue

                    sequence = None
                    qual = None
                    if p_read.is_del:
                        # TODO Not sure about how to estimate quality of deletion?
                        sequence = "-" * (abs(p_read.indel) + 1)
                        qual = p_read.alignment.query_qualities[p_read.query_position_or_next] * (abs(p_read.indel) + 1)
                    elif p_read.indel > 0:
                        # p_read.indel peeks to next CIGAR and determines whether the base FOLLOWING this one is an insertion or not
                        sequence = p_read.alignment.query_sequence[p_read.query_position : p_read.query_position + p_read.indel + 1]
                        qual = p_read.alignment.query_qualities[p_read.query_position : p_read.query_position + p_read.indel + 1]
                    else:
                        sequence = p_read.alignment.query_sequence[p_read.query_position]
                        qual = p_read.alignment.query_qualities[p_read.query_position]

                    if not sequence:
                        print("[WARN] Sequence data seems to not be correctly salvaged from read %s" % p_read.alignment.query_name)
                        continue

                    if curr_read_name not in reads:
                        reads[curr_read_name] = {
                            "rank": np.sum(vcf_handler["region"][1 : LEFTMOST_1pos]), # non-inclusive 1pos end
                            "seq": [],
                            "quals": [],
                            "refs_1pos": [],
                            "read_variants_0pos": [],
                        }
                        if p_read.alignment.query_name in debug_reads:
                            dreads.add(curr_read_name)
                    reads[curr_read_name]["seq"].append(sequence)
                    reads[curr_read_name]["quals"].append(qual)
                    reads[curr_read_name]["refs_1pos"].append(p_col.reference_pos+1)
                    reads[curr_read_name]["read_variants_0pos"].append(p_read.query_position)


            for dread in sorted(dreads):
                r = reads[dread]
                if r:
                    for snp_i, ref_pos in enumerate(r["refs_1pos"]):
                        print (dread, ref_pos, r["seq"][snp_i])
                    print ("RANK", dread, r["rank"])

            if debug_pos:
                for read in reads:
                    for d_pos in set(reads[read]["refs_1pos"]) & debug_pos:
                        i = reads[read]["refs_1pos"].index(d_pos)
                        print (read, d_pos, reads[read]["seq"][i])


            num_reads = len(reads)
            for qi, qname in enumerate(reads):
                progress_q.put({"pos": num_reads-(qi+1), "worker_i": worker_i})

                if not len(reads[qname]["seq"]) > 1:
                    # Ignore reads without evidence
                    continue
                slices += 1

                rank = reads[qname]["rank"]
                support_len = len(reads[qname]["seq"])

                support_seq = "".join([b[0] for b in reads[qname]["seq"]]) # b[0] has the affect of capturing the base before any insertion
                covered_snps += len(support_seq.replace("N", "").replace("_", ""))

                # For each position in the supporting sequence (that is, each covered SNP)
                for i in range(0, support_len):
                    snp_a = support_seq[i]

                    #if support_len == 1:
                    #    if rank == 0:
                    #        hansel.add_observation('_', snp_a, 0, 1)
                    #        hansel.add_observation(snp_a, '_', 1, 2)
                    #    else:
                    #        hansel.add_observation(snp_a, '_', rank+1, rank+2)


                    # For each position in the supporting sequence following i
                    for j in range(i+1, support_len):
                        snp_b = support_seq[j]

                        # Ignore observations who are from an invalid transition
                        if snp_a in ['_', 'N']:
                            continue

                        # Sentinel->A
                        if i==0 and j==1 and rank==0:
                            # If this is the first position in the support (support_pos == 0)
                            # and rank > 0 (that is, this is not the first SNP)
                            # and SNPs a, b are adjacent
                            hansel.add_observation('_', snp_a, 0, 1)
                            hansel.add_observation(snp_a, snp_b, 1, 2)
                            crumbs += 1

                        # B->Sentinel
                        elif (j+rank+1) == vcf_handler["N"] and abs(i-j)==1:
                            # Last observation (abs(i-j)==1),
                            # that ends on the final SNP (j+rank+1 == N)
                            hansel.add_observation(snp_a, snp_b, vcf_handler["N"]-1, vcf_handler["N"])
                            hansel.add_observation(snp_b, '_', vcf_handler["N"], vcf_handler["N"]+1)
                            crumbs += 1

                        # A regular observation (A->B)
                        else:
                            hansel.add_observation(snp_a, snp_b, i+rank+1, j+rank+1)
                            crumbs += 1

                            if use_end_sentinels:
                                if j==(support_len-1) and abs(i-j)==1:
                                    # The last SNP on a read, needs a sentinel afterward
                                    hansel.add_observation(snp_b, '_', j+rank+1, j+rank+2)

    bam_queue = Queue()
    progress_queue = Queue()

    # Queue the wokers
    # TODO Evenly divide, but in future, consider the distn
    # TODO Also consider in general block0 has more work to do
    window_l = round((end_pos - start_pos) / float(n_threads))
    for window_i, window_pos in enumerate(range(start_pos, end_pos+1, window_l)):
        bam_queue.put({
            "start": window_pos,
            "end": window_pos + window_l - 1, # add -1 to stop end of window colliding with next window
            "i": window_i,
            "region_end": end_pos,
        })

    processes = []
    for _ in range(n_threads):
        p = Process(target=bam_worker,
                    args=(bam_queue, progress_queue, _))
        processes.append(p)

    # ...and a progress process
    n_reads = Value('i', 0)
    n_observations = Value('i', 0)
    total_covered_snps = Value('i', 0)
    p = Process(target=progress_worker,
                args=(progress_queue, n_threads, n_reads, total_covered_snps, n_observations))
    processes.append(p)

    for p in processes:
        p.start()

    # Add sentinels
    for _ in range(n_threads):
        bam_queue.put(None)

    # Wait for processes to complete work
    for p in processes:
        p.join()


    hansel.n_slices = n_reads.value
    hansel.n_crumbs = n_observations.value
    sys.stderr.write("[NOTE] Loaded %d breadcrumbs from %d bread slices.\n" % (hansel.n_crumbs, hansel.n_slices))

    hansel.L = int(ceil(float(total_covered_snps.value)/n_reads.value))
    sys.stderr.write("[NOTE] Setting Gretel.L to %d\n" % hansel.L)
    return hansel

def load_fasta(fa_path):
    """
    A convenient wrapper function for constructing a :py:class:`pysam.FastaFile`

    Parameters
    ----------
    fa_path : str
        Path to FASTA

    Returns
    -------

    FASTA File Interface : :py:class:`pysam.FastaFile`
    """
    return pysam.FastaFile(fa_path)


def process_vcf(vcf_path, contig_name, start_pos, end_pos):
    """
    Parse a VCF to extract the genomic positions of called variants.

    Parameters
    ----------
    vcf_path : str
        Path to the VCF file.

    contig_name : str
        Name of the target contig on which variants were called.

    start_pos : int
        The 1-indexed genomic position from which to begin considering variants.

    end_pos : int
        The 1-indexed genomic position at which to stop considering variants.

    Returns
    -------
    Gretel Metastructure : dict
        A collection of structures used for the execution of Gretel.
        The currently used keys are:
            N : int
                The number of observed SNPs
            snp_fwd : dict{int, int}
                A reverse lookup from the n'th variant, to its genomic position on the contig
            snp_rev : dict{int, int}
                A forward lookup to translate the n'th genomic position to its i'th SNP rank
            region : list{int}
                A masked representation of the target contig, positive values are variant positions
    """

    # Open the VCF
    fp = open(vcf_path, 'rb') # assumes bgzip and tabix
    vcf_records = vcf.Reader(fp)
    n_snps = 0
    snp_reverse = {}
    snp_forward = {}
    region = np.zeros(end_pos + 1, dtype=int)
    i = 0
    for record in vcf_records.fetch(contig_name, 0, end_pos): # [0, 1end)
        # record.POS is 1-indexed
        if record.POS < start_pos:
            continue
        if record.POS > end_pos:
            continue

        n_snps += 1
        region[record.POS] = 1
        snp_reverse[i] = record.POS
        snp_forward[record.POS] = i
        i += 1
    fp.close()

    return {
        "N": n_snps,
        "snp_fwd": snp_forward,
        "snp_rev": snp_reverse,
        "region": region,
    }


gretel/cmd.py
import argparse
import numpy as np
import sys
import os

from . import gretel
from . import util

__version__ = "0.0.94"

def main():
    """Gretel: A metagenomic haplotyper."""
    parser = argparse.ArgumentParser(description="Gretel: A metagenomic haplotyper.")
    parser.add_argument("bam")
    parser.add_argument("vcf")
    parser.add_argument("contig")
    parser.add_argument("-s", "--start", type=int, default=1, help="1-indexed included start base position [default: 1]")
    parser.add_argument("-e", "--end", type=int, default=-1, help="1-indexed inlcuded end base position [default: reference length]")

    #parser.add_argument("-l", "--lorder", type=int, default=0, help="Order of markov chain to predict next nucleotide [default calculated from read data]")
    parser.add_argument("-p", "--paths", type=int, default=100, help="Maximum number of paths to generate [default:100]")

    parser.add_argument("--master", default=None, help="Master sequence (will be used to fill in homogeneous gaps in haplotypes, otherwise --gapchar)") #TODO Use something other than N? Should probably be a valid IUPAC
    parser.add_argument("--gapchar", default="N", help="Character to fill homogeneous gaps in haplotypes if no --master [default N]")
    parser.add_argument("--delchar", default="", help="Character to output in haplotype for deletion (eg. -) [default is blank]")

    parser.add_argument("--quiet", default=False, action='store_true', help="Don't output anything other than a single summary line.")
    #parser.add_argument("--sentinels", default=False, action='store_true', help="Add additional sentinels for read ends [default:False][EXPERIMENTAL]")
    parser.add_argument("-o", "--out", default=".", help="Output directory [default .]")
    parser.add_argument("-@", "--threads", type=int, default=1, help="Number of BAM iterators [default 1]")

    parser.add_argument("--debugreads", type=str, default="", help="A newline delimited list of read names to output debug data when parsing the BAM")
    parser.add_argument("--debugpos", type=str, default="", help="A newline delimited list of 1-indexed genomic positions to output debug data when parsing the BAM")
    parser.add_argument("--debughpos", type=str, default=",", help="A comma delimited list of 1-indexed SNP positions to output debug data when predicting haplotypes")

    parser.add_argument("--dumpmatrix", type=str, default=None, help="Location to dump the Hansel matrix to disk")
    parser.add_argument("--dumpsnps", type=str, default=None, help="Location to dump the SNP positions to disk")

    parser.add_argument("--pepper", action="store_true", help="enable a more permissive pileup by setting the pysam pileup stepper to 'all', instead of 'samtools'.\nNote that this will allow improper pairs.")

    parser.add_argument("--version", action="version", version="%(prog)s " + __version__)

    ARGS = parser.parse_args()

    debug_hpos = []
    if ARGS.debughpos:
        for x in ARGS.debughpos.split(","):
            try:
                debug_hpos.append( int(x) )
            except:
                pass

    if ARGS.end == -1:
        ARGS.end = util.get_ref_len_from_bam(ARGS.bam, ARGS.contig)
        sys.stderr.write("[NOTE] Setting end_pos to %d" % ARGS.end)

    debug_reads = set([])
    if ARGS.debugreads:
        debug_fofn = open(ARGS.debugreads)
        for line in debug_fofn:
            debug_reads.add(line.strip())

    debug_pos = set([])
    if ARGS.debugpos:
        debug_fofn = open(ARGS.debugpos)
        for line in debug_fofn:
            debug_pos.add(int(line.strip()))

    VCF_h = util.process_vcf(ARGS.vcf, ARGS.contig, ARGS.start, ARGS.end)
    if ARGS.dumpsnps:
        snp_fh = open(ARGS.dumpsnps, 'w')
        for k in sorted(VCF_h["snp_fwd"].keys()):
            snp_fh.write("%d\t%d\t%d\n" % (VCF_h["snp_fwd"][k]+1, k, k-ARGS.start+1))
        snp_fh.close()

    # Could we optimise for lower triangle by collapsing one of the dimensions
    # such that Z[m][n][i][j] == Z[m][n][i + ((j-1)*(j))/2]
    hansel = util.load_from_bam(ARGS.bam, ARGS.contig, ARGS.start, ARGS.end, VCF_h, n_threads=ARGS.threads, debug_reads=debug_reads, debug_pos=debug_pos, stepper="all" if ARGS.pepper else "samtools")
    original_hansel = hansel.copy()

    if ARGS.dumpmatrix:
        hansel.save_hansel_dump(ARGS.dumpmatrix)

    # Check if there is a gap in the matrix
    for i in range(0, VCF_h["N"]+1):
        marginal = hansel.get_counts_at(i)

        if i > 0:
            snp_rev = VCF_h["snp_rev"][i-1]
        else:
            snp_rev = 0
        if marginal.get("total", 0) == 0:
            sys.stderr.write('''[FAIL] Unable to recover pairwise evidence concerning SNP #%d at position %d
       Gretel needs every SNP to appear on a read with at least one other SNP, at least once.
       There is no read in your data set that bridges SNP #%d with any of its neighbours.

       * If you are trying to run Gretel along an entire contig or genome, please note that
       this is not the recommended usage for Gretel, as it was intended to uncover the
       variation in a metahaplome: the set of haplotypes for a specific gene.
           See our pre-print https://doi.org/10.1101/223404 for more information

       Consider running a prediction tool such as `prokka` on your assembly or reference
       and using the CDS regions in the GFF for corresponding genes of interest to
       uncover haplotypes with Gretel instead.

       * If you are already doing this, consider calling for SNPs more aggressively.
       We use `snpper` (https://github.com/SamStudio8/gretel-test/blob/master/snpper.py)
       to determine any site in a BAM that has at least one read in disagreement with
       the reference as a SNP. Although this introduces noise from alignment and sequence
       error, Gretel is fairly robust. Importantly, this naive calling method will
       likely close gaps between SNPs and permit recovery.

       * Finally, consider that the gaps are indicative that your reads do not support
       one or more parts of your assembly or reference. You could try and find or construct
       a more suitable reference, or reduce the size of the recovery window.

       Sorry :(\n''' % (i, snp_rev, i))
            sys.exit(1)

    PATHS = {}

    # Spew out exciting information about the SNPs
    if not ARGS.quiet:
        print ("i\tpos\tgap\tA\tC\tG\tT\tN\t-\t_\ttot")
        last_rev = 0
        for i in range(0, VCF_h["N"]+1):
            marginal = hansel.get_counts_at(i)
            marginal = {str(x): marginal[x] for x in marginal}
            snp_rev = 0
            if i > 0:
                snp_rev = VCF_h["snp_rev"][i-1]
            print ("%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d" % (
                i,
                snp_rev,
                snp_rev - last_rev,
                marginal.get("A", 0),
                marginal.get("C", 0),
                marginal.get("G", 0),
                marginal.get("T", 0),
                marginal.get("N", 0),
                marginal.get("-", 0),
                marginal.get("_", 0),
                marginal.get("total", 0),
            ))
            last_rev = snp_rev


    # Make some genes
    SPINS = ARGS.paths
    ongoing_mag = 0
    for i in range(0, SPINS):
        init_path, init_prob, init_min = gretel.generate_path(VCF_h["N"], hansel, original_hansel, debug_hpos=debug_hpos)
        if init_path == None:
            break
        current_path = init_path

        MIN_REMOVE = 0.01 # 1%
        if init_min < MIN_REMOVE:
            sys.stderr.write("[RWGT] Ratio %.10f too small, adjusting to %.3f\n" % (init_min, MIN_REMOVE))
            init_min = MIN_REMOVE
        rw_magnitude = gretel.reweight_hansel_from_path(hansel, init_path, init_min)

        #TODO Horribly inefficient.
        current_path_str = "".join([str(x) for x in current_path])
        if current_path_str not in PATHS:
            PATHS[current_path_str] = {
                "hp_current": [],
                "hp_original": [],
                "i": [],
                "i_0": i,
                "n": 0,
                "magnitude": 0,
                "hansel_path": current_path,
            }
        PATHS[current_path_str]["n"] += 1
        PATHS[current_path_str]["i"].append(i)
        PATHS[current_path_str]["magnitude"] += rw_magnitude
        PATHS[current_path_str]["hp_current"].append(init_prob["hp_current"])
        PATHS[current_path_str]["hp_original"].append(init_prob["hp_original"])

    # Output FASTA
    dirn = ARGS.out + "/"
    fasta_out_fh = open(dirn+"out.fasta", "w")
    hfasta_out_fh = open(dirn+"snp.fasta", "w")

    if ARGS.master:
        master_fa = util.load_fasta(ARGS.master)
        master_seq = master_fa.fetch(master_fa.references[0])
    else:
        master_seq = [' '] * ARGS.end

    for p in sorted(PATHS, key=lambda x: PATHS[x]["i_0"]):
        p = PATHS[p]
        path = p["hansel_path"]
        i = p["i_0"]

        seq = list(master_seq[:])
        for j, mallele in enumerate(path[1:]):
            snp_pos_on_master = VCF_h["snp_rev"][j]
            try:
                if mallele == hansel.symbols_d["-"]:
                    # It's a deletion, don't print a SNP
                    seq[snp_pos_on_master-1] = ARGS.delchar
                else:
                    seq[snp_pos_on_master-1] = mallele
            except IndexError:
                print (path, len(seq), snp_pos_on_master-1)
                sys.exit(1)

        # Coerce HanselSymbols to str
        to_write = "".join(str(x) for x in seq[ARGS.start-1 : ARGS.end])
        if not ARGS.master:
            to_write = to_write.replace(' ', ARGS.gapchar)

        fasta_out_fh.write(">%d__%.2f\n" % (i, p["hp_current"][0])) #TODO hp_current or hp_original?
        fasta_out_fh.write("%s\n" % to_write)

        hfasta_out_fh.write(">%d__%.2f\n" % (i, p["hp_current"][0])) #TODO hp_current or hp_original?
        hfasta_out_fh.write("%s\n" % "".join([str(x) for x in path[1:]]))
    fasta_out_fh.close()
    hfasta_out_fh.close()

    #TODO datetime, n_obs, n_slices, avg_obs_len, L, n_paths, n_avg_loglik
    crumb_file = open(dirn+"gretel.crumbs", "w")
    crumb_file.write("# %d\t%d\t%d\t%.2f\n" % (
        VCF_h["N"],
        hansel.n_crumbs,
        hansel.n_slices,
        hansel.L,
    ))

    for p in sorted(PATHS, key=lambda x: PATHS[x]["hp_current"][0], reverse=True):
        p = PATHS[p]
        crumb_file.write("%d\t%d\t%s\t%s\t%.2f\n" % (
                p["i_0"],
                p["n"],
                ",".join(["%.2f" % x for x in p["hp_current"]]),
                ",".join(["%.2f" % x for x in p["hp_original"]]),
                p["magnitude"],
        ))
